src/v/storage/offset_translator_state.cc

```diff
@@ -159,27 +172,68 @@ bool offset_translator_state::add_absolute_delta(
           prev, batch_info{.base_offset = base_offset, .next_delta = delta});
         return true;
     } else {
-        int64_t last_delta = _last_offset2batch.rbegin()->second.next_delta;
-        int64_t gap_length = delta - last_delta;
+        auto [last_offset, last_value] = *_last_offset2batch.rbegin();
+        auto [last_base_offset, last_delta] = last_value;
 
-        if (gap_length != 0) {
-            model::offset last_offset = _last_offset2batch.rbegin()->first;
+        if (prev > last_offset) {
+            int64_t gap_length = delta - last_delta;
             auto base_offset = offset - model::offset{gap_length};
-            vassert(
-              base_offset > last_offset && base_offset < offset,
-              "ntp {}: inconsistent add_absolute_delta (offset {}, delta {}), "
-              "but last_offset: {}, last_delta: {}",
-              _ntp,
-              offset,
-              delta,
+
+            vlog(
+              stlog.info,
+              "add_absolute_delta: last record: key: {} - {{ bo: {}, delta: {} "
+              "}}",
               last_offset,
+              last_base_offset,
               last_delta);
 
+            if (gap_length < 0) {
+                // This may happen if the delta produced by add_gap calls
+                // is not consistent with add_absolute_delta call.
+                throw std::runtime_error(fmt_with_ctx(
+                  fmt::format,
+                  "ntp {}: inconsistent add_absolute_delta (offset {}, delta "
+                  "{}), "
+                  "but last_offset: {}, last_delta: {}, last_base_offset: {}, "
+                  "gap_length: {}",
+                  _ntp,
+                  offset,
+                  delta,
+                  last_offset,
+                  last_delta,
+                  last_base_offset,
+                  gap_length));
+            }
+
             _last_offset2batch.emplace(
               prev,
               batch_info{.base_offset = base_offset, .next_delta = delta});
             return true;
         } else {
+            // The record batch is already added, the delta computed using
+            // previous calls to 'add_gap' and 'add_absolute_delta' should match
+            // the parameter of this call. There is no need to change the state
```

Comment:
So I have a suspicion that we can still end up with inconsistent data in the offset translator even if the segments themselves are good. The problem is that `add_absolute_delta` adds a fake gap, not a real one and if we then go back and start adding real gaps, we'll see inconsistency.

Example (not sure if this is possible with the way `remote_partition` chooses the order of segments, but probably it is):
* execute fetch from offset 300:
* start reading segment [200;400]: do `add_absolute_delta(200, d_200)`
* add some gaps from the segment [200;400]
* segment ends, we request the next segment: it is overlapping segment [100;500]
* start reading segment [100;500]: do `add_absolute_delta(100, d_100)`, but the offset translator state doesn't change! Because offset 100 is before the OT map end.
* start adding gaps from segment [100;500]. Uh-oh, the gaps are inconsistent, the OT state in the 100-200 range is bogus.

I see two solutions:
1. (outlined in another comment) just save the next rp offset and skip batches in the new segment until we reach it
2. make `add_absolute_delta` throw away all OT map entries greater than `offset` - this way if we read some batches twice, that's okay because we will just rebuild OT state. This way we can even leave code of `add_gap` as-is because it will only add batches to the end of the OT map.